File: Problems/CampusBikesII.py
import sys
from typing import List


class Solution:

    def assignBikes(workers: List[List[int]], bikes: List[List[int]]) -> int:

        def dist(x, y):
            return abs(x[0] - y[0]) + abs(x[1] - y[1])

        def calculate(i, visited, cache):
            if i == len(workers):
                return 0
            if (i, visited) in cache:
                return cache[(i, visited)]
            ans = float('inf')
            for j, bike in enumerate(bikes):
                if visited & 1 << j:
                    continue
                ans = min(
                    ans, dist(bike, workers[i]) + calculate(i+1, visited | 1 << j, cache))
            cache[(i, visited)] = ans
            return ans

        return calculate(0, 0, {})

    memo = [-1 for _ in range(1025)]

    def assignBikes2(self, workers: List[List[int]], bikes: List[List[int]]) -> int:
        return self.helper(self, workers, bikes, 0, 0)

    def dist2(self, x, y):
        return abs(x[0] - y[0]) + abs(x[1] - y[1])

    def helper(self, workers, bikes, workerIndex, visited) -> int:
        if workerIndex >= len(workers):
            return 0
        if self.memo[visited] != -1:
            return self.memo[visited]
        res = sys.maxsize
        for i, bike in enumerate(bikes):
            if visited & (1 << i) != 0:
                continue
            res = min(res, self.dist2(self, workers[workerIndex], bike) + self.helper(
                self, workers, bikes, workerIndex + 1, (visited | (1 << i))))
            self.memo[visited] = res
        return res


workers = [[0, 0], [1, 0], [2, 0], [3, 0], [4, 0]]
bikes = [[0, 999], [1, 999], [2, 999], [3, 999], [4, 999]]
print(Solution.assignBikes2(Solution, workers, bikes))

File: Problems/CampusBikesII.cpp
#include <iostream>
#include <vector>
#include <set>
#include <queue>
using namespace std;

class Solution
{
public:
    int dist(vector<int> &worker, vector<int> &bike)
    {
        return abs(worker[0] - bike[0]) + abs(worker[1] - bike[1]);
    }

    int assignBikes(vector<vector<int>> &workers, vector<vector<int>> &bikes)
    {
        priority_queue<pair<int, pair<int, int>>, vector<pair<int, pair<int, int>>>, greater<pair<int, pair<int, int>>>> pq;
        pq.push(make_pair(0, make_pair(0, 0)));
        set<pair<int, int>> set;
        while (true)
        {
            pair<int, pair<int, int>> p = pq.top();
            pq.pop();
            int minCost = p.first;
            int index = p.second.first;
            int mask = p.second.second;
            if (set.find(p.second) != set.end())
                continue;
            if (index == workers.size())
                return minCost;
            for (int i = 0; i < bikes.size(); i++)
            {
                if ((mask & (1 << i)) == 0)
                {
                    pair<int, pair<int, int>> newP;
                    newP.first = minCost + dist(workers[index], bikes[i]);
                    newP.second.first = index + 1;
                    newP.second.second = mask | (1 << i);
                    pq.push(newP);
                }
            }
        }
    }
};